Implement Compact Blocks BIP152

ptschip opened the pull request:
This works but not yet complete, been running on mainnet for a few days now with any issues... Have to add/fix the p2p tests and also update the MSG_THINBLOCK which conflicts with MSG_CMPCT_BLOCK;  I think the best approach here is to upgrade to xthins version 2 using XVERSION, and then also rather than rely on the enum values we'll create a separate p2p message called "get_thin", rather than use the "getdata" message (this way we won't have any future conflicts if some other project decides to add more enum message types.

Furthermore, the download timers are  no longer active when xthin, graphene, and compact blocks are all turned on. But we can still gather download stats for any one particular feature by turning the others off.  So for instance , if you turn off compactblocks and graphene, the download timers will automatically be turned on and you'll get good statistics for xthins. You can of course do the same for cmpct blocks (when i get the instrumentation done) by turning off xthin/graphene.

TheBlueMatt committed:
Add partial-block block encodings API

TheBlueMatt committed:
Add some blockencodings tests

Move blockencodings tests to src/test/compactblocks_tests.cpp

Also added dagurval's validate compact block tests

TheBlueMatt committed:
Add protocol messages for short-ids blocks

TheBlueMatt committed:
Add reconstruction debug logging

sipa committed:
Implement SipHash in Python

sdaftuar committed:
Add support for compactblocks to mininode

Conflicts:
	qa/rpc-tests/test_framework/mininode.py

sdaftuar committed:
Add p2p test for BIP 152 (compact blocks)

dagurval committed:
initialize uint256 to null

ptschip committed:
Add log level CMPCT for compact blocks

ptschip committed:
Remove type and version from serialization ops

ptschip committed:
add blockencodings.cpp/.h to .formatted-files

ptschip committed:
Use MakeTransactionRef in blockencodings.cpp

ptschip committed:
Implement Compact Blocks (BIP 152)

Cherry-pick from dagurval for blockencodings mods

Move MSG_THINBLOCK up in value

MSG_THINBLOCK has a conflict with MSG_CMPCT_BLOCK and since there
are no other implementations that use the MSG_THINBLOCK message type
then we can easily do the upgrade here without much disruption.
Furthermore, an MSG_THINBLOCK is rarely if ever requested and if
one was it would simpley timeout and another block would be requested.

ptschip committed:
Minor text modification to thin timer.

ptschip committed:
request full block if too many hashes to request

ptschip committed:
simplify code for requesting full blocks

ptschip committed:
Turn off compact blocks for sendheaders.py

ptschip committed:
Update IsCompactBlockValid()

ptschip committed:
Tidy up compactblock statitics gathering

and add output for statistics to getnetworkinfo.

ptschip committed:
Introduce BU_XTHIN_VERSION

ptschip committed:
Add request compact block as failover for graphene

ptschip committed:
Add get_thin message request

sickpig commented:
@ptschip 

[formatting](https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/474293070#L783) problem in `allowed_args.cpp`

[requestmanager_tests.cpp](https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/474293070#L1003) unit test failing: 

```
../../src/test/requestmanager_tests.cpp(67): Entering test suite "requestmanager_tests"
../../src/test/requestmanager_tests.cpp(69): Entering test case "blockrequest_tests"
/usr/include/c++/4.8/debug/vector:346:error: attempt to subscript container 
    with out-of-bounds index 61, but container only holds 61 elements.
```

ptschip committed:
Fix a few tests

ptschip committed:
Fix format

Fix rebase

ptschip commented:
@sickpig ok fixed 

> @ptschip
> 
> [formatting](https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/474293070#L783) problem in `allowed_args.cpp`
> 
> [requestmanager_tests.cpp](https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/474293070#L1003) unit test failing:
> 
> ```
> ../../src/test/requestmanager_tests.cpp(67): Entering test suite "requestmanager_tests"
> ../../src/test/requestmanager_tests.cpp(69): Entering test case "blockrequest_tests"
> /usr/include/c++/4.8/debug/vector:346:error: attempt to subscript container 
>     with out-of-bounds index 61, but container only holds 61 elements.
> ```

ptschip committed:
Remove redundant ThinType from ThinTypeRelay class

ptschip committed:
Consolidate thin type object request tracking

Previously we had each block type handle tracking for a different
set of counters. Here we use just one set for all thin block types
and also take the tracking out of CNode and into the ThinTypRelay
tracking object.

ptschip commented:
Taking out of WIP

Greg-Griffith commented:
I have an issue/concern with this PR. Please hold off on merging it until we can discuss it as a group.

Justaphf commented on src/blockrelay/blockrelay_common.cpp:
```diff
@@ -104,23 +107,61 @@ bool ThinTypeRelay::HasBlockRelayTimerExpired(const uint256 &hash)
 
 bool ThinTypeRelay::IsBlockRelayTimerEnabled()
```
Is there a reason this method is written out "long-hand"?  The repeated `else if` cases are all nearly identical, with the only differences being which conditions are NOT'ed, which makes it hard to review and very easy to miss logic errors.

Provided there aren't any intended special caveats I'm missing in the logic, wouldn't it be cleaner to reduce this method down similar to below?  I've revised the comments in the suggested change with my understanding of what this method is supposed to do as well as modifying the code.

NOTE: I also wrote a quick verification program that tests every possible input against the current implementation vs the suggestion below and the results deviate in 5 of 64 cases.  This is because these 5 cases fall through all checks in the original implementation when I don't believe they should.
```suggestion
bool ThinTypeRelay::IsBlockRelayTimerEnabled()
{
    // Only engage the timer if one or more, but not all, thin type relays are active.
    // If all types are active, or all inactive, then we do not need the timer.
    // Generally speaking all types will be active and we can return early.
    if (IsThinBlocksEnabled() && IsGrapheneBlockEnabled() && IsCompactBlocksEnabled())
        return false;
    if (!IsThinBlocksEnabled() && !IsGrapheneBlockEnabled() && !IsCompactBlocksEnabled())
        return false;

    // The thin relay timer is only relevant if we have a specific thin relay type active
    // AND we have peers connected which also support that thin relay type
    bool fThinBlockPossible = IsThinBlocksEnabled() && nThinBlockPeers > 0;
    bool fGraphenePossible = IsGrapheneBlockEnabled() && nGraphenePeers > 0;
    bool fCompactBlockPossible = IsCompactBlocksEnabled() && nCompactBlockPeers > 0;

    return fThinBlockPossible || fGraphenePossible || fCompactBlockPossible;
}
```

ptschip replied:
@Justaphf thanks, will take a look...you're probably right...i think i just wrote it that way because i was writing down each case as i went along to make sure the test cases worked out.  It could probably be condensed now...

ptschip replied:
@Justaphf great, it works...much better ....thankyou!!!  Do you want to submit a PR to my branch and I can merge it in?

Justaphf replied:
Added a PR against your branch.

ptschip replied:
thanks, merged.

ptschip committed:
Fix bug: ABC nodes not receiving compact blocks

ptschip committed:
Use debug log level cmpctblock instead of cmpctblocks

This keeps us consistent with other implementations

Justaphf committed:
Refactor IsBlockRelayTimerEnabled method

ptschip commented:
Ok, thanks, fixed

> @ptschip
> 
> there's a failure related to the new `getlogcategories.py` test.
> 
> Even if such a test is not included in your branch is run nonetheless by Travis cause current dev with this PR on top is used (even if you did not rebase your branch).
> 
> So once you rebase you further need to append `cmpctblock` string to the expected result that is checked by `getlogcategories.py`, i.e.:
> 
> ```c++
> diff --git a/qa/rpc-tests/getlogcategories.py b/qa/rpc-tests/getlogcategories.py
> index 1b9380d..4973822 100755
> --- a/qa/rpc-tests/getlogcategories.py
> +++ b/qa/rpc-tests/getlogcategories.py
> @@ -67,7 +67,7 @@ class GetLogCategories (BitcoinTestFramework):
>          cat7 = self.nodes[7].log()
>          cat8 = self.nodes[8].log()
>  
> -        exp0 = "coindb tor addrman libevent http rpc partitioncheck bench prune reindex mempoolrej blk evict parallel rand req bloom estimatefee lck proxy dbase selectcoins zmq qt ibd respend weakblocks"
> +        exp0 = "coindb tor addrman libevent http rpc partitioncheck bench prune reindex mempoolrej blk evict parallel rand req bloom estimatefee lck proxy dbase selectcoins zmq qt ibd respend weakblocks cmpctblock"
> ```

ptschip committed:
Only send SENDCMPCT message is compact blocks is enabled.

Previoulsy was using the Core method of checking protocol version which
is not useful for us. Futhermore, protocol number here is not really
relevant because it's ok to send the sendcmpct message to a peer
that doesn't necessarily support this feature, since they will just
ignore it.

ptschip commented:
rebased

ptschip committed:
fix rebase

Justaphf committed:
Add compact block statistics to Debug window UI

Justaphf commented:
@ptschip I'm seeing a lot of errors in the log files running this PR indicating too many thin type requests.  Is this expected?
For example:
> 2019-01-31 04:39:01 Misbehaving: xx.xx.xx.xx:8333 (55) (0 -> 50)
> 2019-01-31 04:39:01 ERROR: xx.xx.xx.xx:8333 (55) is misbehaving. Making too many thin type requests.

On a run of about 1.5 hours I had 56 different nodes produce this same pair of error messages.  I spot checked the first 10 or so addresses and they were all BUCash nodes.  Seems like there may be a backwards compatibility issue with older BU nodes.

ptschip committed:
Move thintype request tracking to the request manager.

These object requests should be tracked separately for each
peer.

sickpig commented on qa/rpc-tests/ctor.py:
```diff
@@ -207,6 +207,7 @@ def run_test (self):
         waitFor(15, lambda: self.nodes[2].getblockcount() == ctorForkTipCount)
         waitFor(15, lambda: self.nodes[3].getblockcount() == ctorForkTipCount)
 
+        print ("got here last ")
```
need to be removed

ptschip replied:
thx, done

ptschip replied:
done

ptschip committed:
remove print statement from ctor.py

sickpig commented on src/init.cpp:
```diff
@@ -967,16 +967,17 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
-        LOGA("Using %d message handler threads\n", numMsgHandlerThreads.Value());
     }
+    LOGA("Using %d message handler threads\n", numMsgHandlerThreads.Value());
+
```
nice catch

sickpig commented on src/init.cpp:
```diff
@@ -967,16 +967,17 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
         numTxAdmissionThreads.Set(nThreads);
-        LOGA("Using %d transaction admission threads\n", numTxAdmissionThreads.Value());
     }
+    LOGA("Using %d transaction admission threads\n", numTxAdmissionThreads.Value());
```
ditto

sickpig commented on src/blockrelay/blockrelay_common.cpp:
```diff
@@ -104,26 +106,21 @@ bool ThinTypeRelay::HasBlockRelayTimerExpired(const uint256 &hash)
+    // If all types are active, or all inactive, then we do not need the timer.
+    // Generally speaking all types will be active and we can return early.
+    if (IsThinBlocksEnabled() && IsGrapheneBlockEnabled() && IsCompactBlocksEnabled())
         return false;
```
I've started `bitcoind` using the following flags:

```
-use-thinblocks=1 -use-grapheneblocks=1 -use-compactblocks=1
```

according to my understanding this should have lead to have the preferential timer disabled. 
 
But it seems like that the preferential timer is used in anyway, in fact I'm seeing in the log this kind of statement:

```
2019-01-31 14:24:00 Starting Preferential Block Relay timer (944 millis)                                                                                                                                                                                                          
2019-01-31 14:24:00 Clearing Preferential BlockRelay timer 
```

Am I missing something? 


edit:

looking at the code it seems that the latter statement is triggered even if the timer is disabled. Guess that we should log it only `if ThinTypeRelay::IsBlockRelayTimerEnable()` 

The former should is in `HasBlockRelayTimerExpired` and I guess we should avoid to execute it entirely.

ptschip replied:
ok, fixed the log entries so they don't get printed out if the timer is disabled.

> @ptschip
> 
> > It's just the log statement getting printed out, but I can fix that.
> 
> ok thanks. But I think that `2019-01-31 14:24:00 Starting Preferential Block Relay timer (944 millis)` it is only be produce if `HasBlockRelayTimerExpired()` is executed and I think we could avoid to execute the entire function in case of the timer is disabled.
> 
> > As for @bissias  PR, I think he'll have to update and rebase after this  one gets merged.
> 
> Agreed.

sickpig commented on src/net.cpp:
```diff
@@ -2876,13 +2877,31 @@ CNode::CNode(SOCKET hSocketIn, const CAddress &addrIn, const std::string &addrNa
+    compactBlockWaitingForTxns = -1;
+    shorttxidk0 = 0;
+    shorttxidk1 = 0;
+
```
@ptschip @bissias both this PR and #1551 implement `GetShortID()`, wonder if we can consolidate the code in one place only.

bissias replied:
@sickpig I thought about doing the same. But I stuffed a version check into my implementation of `GetShortID` (which makes the rest of the code a little cleaner). I also thought initially about creating a class for SipHashing cheap hashes, but there was so little code it didn't seam to be worth it. At any rate, I'm happy to refactor as you all see fit.

ptschip commented:
It's just the log statement getting printed out, but I can fix that.

As for @bissia PR, I think he'll have to update and rebase after this 
one gets merged.

On 31/01/2019 7:01 a.m., Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/blockrelay_common.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r252691409>:
>
> > @@ -104,26 +106,21 @@ bool ThinTypeRelay::HasBlockRelayTimerExpired(const uint256 &hash)
>  
>  bool ThinTypeRelay::IsBlockRelayTimerEnabled()
>  {
> -    // Only engage the timer if at least one thin type relay is active.
> -    if (!IsThinBlocksEnabled() && !IsGrapheneBlockEnabled())
> +    // Only engage the timer if one or more, but not all, thin type relays are active.
> +    // If all types are active, or all inactive, then we do not need the timer.
> +    // Generally speaking all types will be active and we can return early.
> +    if (IsThinBlocksEnabled() && IsGrapheneBlockEnabled() && IsCompactBlocksEnabled())
>          return false;
>
> I've started |bitcoind| using the following flags:
>
> |-use-thinblocks=1 -use-grapheneblocks=1 -use-compactblocks=1 |
>
> according to my understanding this should have lead to have the
> preferential timer disabled.
>
> But it seems like that the preferential timer is used in anyway, in
> fact I'm seeing in the log this kind of statement:
>
> |2019-01-31 14:24:00 Starting Preferential Block Relay timer (944
> millis) 2019-01-31 14:24:00 Clearing Preferential BlockRelay timer |
>
> Am I missing something?
>
> edit:
>
> looking at the code it seems that the latter statement is triggered
> even if the timer is disabled. Guess that we should log it only |if
> ThinTypeRelay::IsBlockRelayTimerEnable()|
>
> The former should is in |HasBlockRelayTimerExpired| and I guess we
> should avoid to execute it entirely.
>
> ------------------------------------------------------------------------
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r252699748>:
>
> > +    thinBlockWaitingForTxns = -1;
> +    nXthinBloomfilterSize = 0;
> +    addrFromPort = 0;
> +
> +    // graphene
> +    nLocalGrapheneBlockBytes = 0;
> +    nSizeGrapheneBlock = 0;
> +    grapheneBlockWaitingForTxns = -1;
> +
> +    // compact blocks
> +    nLocalCompactBlockBytes = 0;
> +    nSizeCompactBlock = 0;
> +    compactBlockWaitingForTxns = -1;
> +    shorttxidk0 = 0;
> +    shorttxidk1 = 0;
> +
>
> @ptschip <https://github.com/ptschip> @bissias
> <https://github.com/bissias> both this PR and #1551
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1551>
> implement |GetShortID()|, wonder if we can consolidate the code in one
> place only.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-198622332>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0LZDBWtxousgkyYa80BQJ3zdHoYQks5vIwU-gaJpZM4Zlisx>.
>

sickpig commented:
@ptschip 

> It's just the log statement getting printed out, but I can fix that.

ok thanks. But I think that `2019-01-31 14:24:00 Starting Preferential Block Relay timer (944 millis)` it is only be produce if `HasBlockRelayTimerExpired()` is executed and I think we could avoid to execute the entire function in case of the timer is disabled. 

> As for @bissias  PR, I think he'll have to update and rebase after this  one gets merged.

Agreed.

sickpig commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1216 @@
+    pnode->vShortCompactBlockHashes.clear();
+    pnode->mapMissingTx.clear();
+
+    LOG(CMPCT, "Total in memory compactblockbytes size after clearing a compactblock is %ld bytes\n",
```
missing space between compactblock and size?

ptschip replied:
done

sickpig commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1216 @@
+            }
+        }
+    } // End locking orphanpool.cs, mempool.cs and cs_xval
+    LOG(CMPCT, "Total in memory compactblockbytes size is %ld bytes\n", compactdata.GetCompactBlockBytes());
```
same here

ptschip replied:
done

ptschip committed:
A variety of nits

Mostly this is about preventing the log entries that the thintimer
was started and then stopped when in fact the timer was disabled
and there was no necessity for any log statements.

Fix an unrelated nit with pushKV() in unlimited.cpp. This was
causing a problem when compiling in debug.

A couple of other log nits

Justaphf committed:
Reorder getnetworkinfo order for compact stats and update help text for missing outputs

Justaphf commented:
@ptschip looking at the logs for running the past 
~9.5 hours I don't see a single one of those "too many thin type requests" errors.

Interestingly, the statistics are showing 100% of the last 50 blocks received via compact and 100% sent by XThin or Graphene (nearly a 50/50 split, between those for sending).  I guess this is indicative of ABCs node dominance and that most miners are submitting blocks with ABC client.

ptschip commented:
@Justaphf merged your UI changes.

gandrewstone commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1216 @@
+        if (cmpctblock.prefilledtxn[i].tx.IsNull())
+            throw std::invalid_argument("null tx in compact block");
+
+        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
```
does this mean compact blocks are only good for 65k tx?

sickpig replied:
Now that I think about @dagur modified xt and ABC code so that it uses an uint32 (or 64) for index

https://github.com/bitcoinxt/bitcoinxt/pull/345
https://reviews.bitcoinabc.org/D1146

Dagur replied:
Seriously guys, stop tagging me in this project. I have nothing to do with it

sickpig replied:
sorry @Dagur, I meant to tag @dagurval.

gandrewstone commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1216 @@
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Compact block message received from a non compactblock node, peer=%s", pfrom->GetLogName());
+    }
```
Is this check needed?  An attack would simply mark itself of compact block capable.  If there is no strong reason for this check, we should be tolerant in what we receive.

ptschip commented:
On 04/02/2019 12:31 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253634371>:
>
> > +{
> +    return ::GetShortID(shorttxidk0, shorttxidk1, txhash);
> +}
> +
> +void validateCompactBlock(const CompactBlock &cmpctblock)
> +{
> +    if (cmpctblock.header.IsNull() || (cmpctblock.shorttxids.empty() && cmpctblock.prefilledtxn.empty()))
> +        throw std::invalid_argument("empty data in compact block");
> +
> +    int32_t lastprefilledindex = -1;
> +    for (size_t i = 0; i < cmpctblock.prefilledtxn.size(); i++)
> +    {
> +        if (cmpctblock.prefilledtxn[i].tx.IsNull())
> +            throw std::invalid_argument("null tx in compact block");
> +
> +        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
>
> does this mean compact blocks are only good for 65k tx?
>
this is for prefilled txns only...in the general case there will be only
1 prefilled (the coinbase)...in compact blocks high bandwidth they may
prefill more

txns, (but we havn't implemented compact blocks 2)

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-199792872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Fa49wZ9NQCDKdtg3ni2G1MMIuxRks5vKJi-gaJpZM4Zlisx>.
>

ptschip commented:
actually now i'm not sure...i'll have to review this again.  This is
straight out of XT's code which is probably straight out of Core.  But
the way they calculate these index values is not very intuitive (from
what i remember) and I'll have to refresh my memory first.

On 04/02/2019 12:31 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253634371>:
>
> > +{
> +    return ::GetShortID(shorttxidk0, shorttxidk1, txhash);
> +}
> +
> +void validateCompactBlock(const CompactBlock &cmpctblock)
> +{
> +    if (cmpctblock.header.IsNull() || (cmpctblock.shorttxids.empty() && cmpctblock.prefilledtxn.empty()))
> +        throw std::invalid_argument("empty data in compact block");
> +
> +    int32_t lastprefilledindex = -1;
> +    for (size_t i = 0; i < cmpctblock.prefilledtxn.size(); i++)
> +    {
> +        if (cmpctblock.prefilledtxn[i].tx.IsNull())
> +            throw std::invalid_argument("null tx in compact block");
> +
> +        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
>
> does this mean compact blocks are only good for 65k tx?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-199792872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Fa49wZ9NQCDKdtg3ni2G1MMIuxRks5vKJi-gaJpZM4Zlisx>.
>

ptschip commented:
On 04/02/2019 1:12 p.m., Dagur Páll Ammendrup wrote:
>
> *@Dagur* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253647792>:
>
> > +{
> +    return ::GetShortID(shorttxidk0, shorttxidk1, txhash);
> +}
> +
> +void validateCompactBlock(const CompactBlock &cmpctblock)
> +{
> +    if (cmpctblock.header.IsNull() || (cmpctblock.shorttxids.empty() && cmpctblock.prefilledtxn.empty()))
> +        throw std::invalid_argument("empty data in compact block");
> +
> +    int32_t lastprefilledindex = -1;
> +    for (size_t i = 0; i < cmpctblock.prefilledtxn.size(); i++)
> +    {
> +        if (cmpctblock.prefilledtxn[i].tx.IsNull())
> +            throw std::invalid_argument("null tx in compact block");
> +
> +        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
>
> Seriously guys, stop tagging me in this project. I have nothing to do
> with it
>
@dagur sorry there's somebody here named @dagurval , easy mistake to make.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253647792>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0JkMNpj5vkra1sbeHP-mawTzxPLDks5vKKI7gaJpZM4Zlisx>.
>

ptschip commented:
I don't feel strongly about it so I'll take it out, also will take it
out of thinblocks and grapheneblocks

On 04/02/2019 12:34 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253635179>:
>
> > +    }
> +}
> +
> +/**
> + * Handle an incoming compactblock.  The block is fully validated, and if any transactions are missing, we fall
> + * back to requesting a full block.
> + */
> +
> +
> +bool CompactBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom)
> +{
> +    if (!pfrom->CompactBlockCapable())
> +    {
> +        dosMan.Misbehaving(pfrom, 100);
> +        return error("Compact block message received from a non compactblock node, peer=%s", pfrom->GetLogName());
> +    }
>
> Is this check needed? An attack would simply mark itself of compact
> block capable. If there is no strong reason for this check, we should
> be tolerant in what we receive.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-199793850>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0JqApIFukAC3pxbp5WLS9bFK-P51ks5vKJlLgaJpZM4Zlisx>.
>

ptschip committed:
Remove checks graphene, thinblock or compactblock capable

There is little benefit to checking this at the beginning of handling
p2p messages because an attacker can easily fake this state.

gandrewstone commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1198 @@
+    {
+        CBlock block;
+        const Consensus::Params &consensusParams = Params().GetConsensus();
+        if (!ReadBlockFromDisk(block, hdr, consensusParams))
```
before reading the block from disk, we should make sure its near the tip and perhaps limit the number of rerequests per 10 minutes per node.

gandrewstone commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1198 @@
+    CBlockIndex *hdr = LookupBlockIndex(inv.hash);
+    if (!hdr)
+    {
+        dosMan.Misbehaving(pfrom, 20);
```
could this happen during a reorg?

ptschip commented:
I dont' see how...it's not looking for the tip but rather an entry in
the blockindex , and since they never get deleted , if the entry is
there is will return a value regardless of re-org or not.

On 04/02/2019 7:16 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253724937>:
>
> > +    if (compactReRequest.indexes.empty() || compactReRequest.blockhash.IsNull())
> +    {
> +        dosMan.Misbehaving(pfrom, 100);
> +        return error("incorrectly constructed getblocktxn received.  Banning peer=%s", pfrom->GetLogName());
> +    }
> +
> +    // We use MSG_TX here even though we refer to blockhash because we need to track
> +    // how many xblocktx requests we make in case of DOS
> +    CInv inv(MSG_TX, compactReRequest.blockhash);
> +    LOG(CMPCT, "received getblocktxn for %s peer=%s\n", inv.hash.ToString(), pfrom->GetLogName());
> +
> +    std::vector<CTransaction> vTx;
> +    CBlockIndex *hdr = LookupBlockIndex(inv.hash);
> +    if (!hdr)
> +    {
> +        dosMan.Misbehaving(pfrom, 20);
>
> could this happen during a reorg?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-199903804>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0NpRaDgPxFh2hWMZ-tnNXw-VrSOzks5vKPd7gaJpZM4Zlisx>.
>

ptschip commented:
we do rate limit there already, but it's done in a prior step in
net_processing.cpp just before we call HandleMessage().  It's the
requester.CheckForRequestDOS().

But, I'll add preventing requesting far from the tip.


On 04/02/2019 7:15 p.m., Andrew Stone wrote:

> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253724860>:
>
> > +    // how many xblocktx requests we make in case of DOS
> +    CInv inv(MSG_TX, compactReRequest.blockhash);
> +    LOG(CMPCT, "received getblocktxn for %s peer=%s\n", inv.hash.ToString(), pfrom->GetLogName());
> +
> +    std::vector<CTransaction> vTx;
> +    CBlockIndex *hdr = LookupBlockIndex(inv.hash);
> +    if (!hdr)
> +    {
> +        dosMan.Misbehaving(pfrom, 20);
> +        return error("Requested block is not available");
> +    }
> +    else
> +    {
> +        CBlock block;
> +        const Consensus::Params &consensusParams = Params().GetConsensus();
> +        if (!ReadBlockFromDisk(block, hdr, consensusParams))
>
> before reading the block from disk, we should make sure its near the
> tip and perhaps limit the number of rerequests per 10 minutes per node.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-199903731>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0EcNqsgnO1BvbFJw4GM4ZPTRqH2-ks5vKPdYgaJpZM4Zlisx>.
>

ptschip committed:
Prevent requesting get_blocktx, get_grblocktx and getblocktxn too far from tip

We limit how far away from the tip you can request by using the same
value DEFAULT_BLOCKS_FROM_TIP that we now use for testing if
the chain is nearly synced in IsChainNearlySyncd().  This way an
attacker can not keep our disk busy by forcing us to load blocks from
any part of the disk but will be restricted to just a few blocks which
would likely still be in cache if re-requested over and over.

ptschip commented:
Ok refreshed my memory.  The answer is Yes and No...this would only
effect compact blocks high bandwidth mode which we have not implemented.
And even then it may not entirely restrict it because the index is
calculated from the last prefilled txn...it's a bit odd but that's how
they implemented it presumably to save a couple of bytes per pre-filled txn.

04/02/2019 12:31 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253634371>:
>
> > +{
> +    return ::GetShortID(shorttxidk0, shorttxidk1, txhash);
> +}
> +
> +void validateCompactBlock(const CompactBlock &cmpctblock)
> +{
> +    if (cmpctblock.header.IsNull() || (cmpctblock.shorttxids.empty() && cmpctblock.prefilledtxn.empty()))
> +        throw std::invalid_argument("empty data in compact block");
> +
> +    int32_t lastprefilledindex = -1;
> +    for (size_t i = 0; i < cmpctblock.prefilledtxn.size(); i++)
> +    {
> +        if (cmpctblock.prefilledtxn[i].tx.IsNull())
> +            throw std::invalid_argument("null tx in compact block");
> +
> +        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
>
> does this mean compact blocks are only good for 65k tx?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-199792872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Fa49wZ9NQCDKdtg3ni2G1MMIuxRks5vKJi-gaJpZM4Zlisx>.
>

ptschip committed:
Remove the clearing of thintype data in parallel.cpp

This clearing is redundant since it's already done at the beginning
of receiving a thintype block but also it can cause us to delete
data which we still need for a thintype block which is processing
still but a full block has completed. It's a rare event but has happened
in some of our python tests.

gandrewstone commented on src/blockrelay/compactblock.cpp:
```diff
@@ -0,0 +1,1198 @@
+    // Get the full hashes from the compactReReqResponse and add them to the compactBlockHashes vector.  These should
+    // be all the missing or null hashes that we re-requested.
+    int count = 0;
+    for (size_t i = 0; i < pfrom->vCompactBlockHashes.size(); i++)
```
since we are iterating through vCompactBlockHashes.size() but in line 516 referencing pfrom->vShortCompactBlockHashes[i], let's DbgAssert(vCompactBlockHashes.size() == pfrom->vShortCompactBlockHashes.size(), ...) here

ptschip commented:
done

On 05/02/2019 5:25 a.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#discussion_r253860523>:
>
> > +        compactdata.ClearCompactBlockData(pfrom, inv.hash);
> +
> +        LOG(CMPCT,
> +            "Received compactReReqResponse but returning because we already have this block %s on disk, peer=%s\n",
> +            inv.hash.ToString(), pfrom->GetLogName());
> +        return true;
> +    }
> +
> +    // Create the mapMissingTx from all the supplied tx's in the compactblock
> +    for (const CTransaction &tx : compactReReqResponse.txn)
> +        pfrom->mapMissingTx[GetShortID(pfrom->shorttxidk0, pfrom->shorttxidk1, tx.GetHash())] = MakeTransactionRef(tx);
> +
> +    // Get the full hashes from the compactReReqResponse and add them to the compactBlockHashes vector.  These should
> +    // be all the missing or null hashes that we re-requested.
> +    int count = 0;
> +    for (size_t i = 0; i < pfrom->vCompactBlockHashes.size(); i++)
>
> since we are iterating through vCompactBlockHashes.size() but in line
> 516 referencing pfrom->vShortCompactBlockHashes[i], let's
> DbgAssert(vCompactBlockHashes.size() ==
> pfrom->vShortCompactBlockHashes.size(), ...) here
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1543#pullrequestreview-200070131>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0GBwI843GQDv1ShdQNxo0F4AfoKzks5vKYY_gaJpZM4Zlisx>.
>

ptschip committed:
Add DbgAssert in compactblocks

Ensure that the maps for the short hashes and full hashes match
before continuing.

ptschip committed:
Support larger blocks for compact blocks

Previously the index was a uint16_t which supported a very small
max block size of perhaps a few megabytes. Using a uint32_t for this
index gives us support up to approx 1TB blocks.

gandrewstone commented on src/net_processing.cpp:
```diff
@@ -258,7 +268,7 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
             // priority messages and we don't want to sit here processing a large number of messages
             // while we hold the cs_main lock, but rather allow these messages to be sent first and
             // process the return message before potentially reading from the queue again.
-            if (inv.type == MSG_BLOCK || inv.type == MSG_FILTERED_BLOCK || inv.type == MSG_THINBLOCK)
```
don't we also want high priority handling of graphene and thin blocks?

ptschip replied:
This section of code above only refers to getdata requests.  graphene and thinblocks are not and can not be requested by a getdata...

ptschip replied:
or i should say, only the legacy THINBLOCK requests can be still requested by getdata, but we've had to move away from that for compact blocks.  Perhaps it would be more clear if I swapped the MSG_THINBLOCK for MSG_COMPACT_BLOCK...I'll do that.

ptschip replied:
oh nvm...it's already there on the next line.

bissias committed:
clear blocks in-flight instead of block data

bissias committed:
remove redundant call to ClearThinTypeBlockInFlight

ptschip commented:
rebased
